src/engine/model_runner.cpp

```diff
@@ -0,0 +1,163 @@
+
+    InputParameters params;
+    params.empty_kv_cache = false;
+    params.num_sequences = batch_size;
+    params.q_max_seq_len = 1;
+    params.kv_max_seq_len = k_max_seq_len;
+    params.q_cu_seq_lens = q_cu_seq_lens.slice(
+        /*dim=*/0, /*start=*/0, /*end=*/batch_size + 1);
+    params.kv_cu_seq_lens = kv_cu_seq_lens.slice(
+        /*dim=*/0, /*start=*/0, /*end=*/batch_size + 1);
+    params.new_cache_slots = new_cache_slots.slice(
+        /*dim=*/0, /*start=*/0, /*end=*/batch_size);
+    params.block_tables = block_tables.slice(
+        /*dim=*/0, /*start=*/0, /*end=*/batch_size);
+
+    // capture graph
+    graph->capture(model_, flatten_tokens, flatten_positions, params, kv_cache);
+
+    // save the graph
+    graphs_[batch_size] = std::move(graph);
+  }
+  LOG(INFO) << "Finished capturing CUDA graphs";
+}
+
+// tokens: [num_tokens]
+// positions: [num_tokens] token pos in the sequence
+// returns: [num_tokens, hidden_size]
+torch::Tensor ModelRunner::forward(const torch::Tensor& tokens,
+                                   const torch::Tensor& positions,
+                                   std::vector<KVCache>& kv_caches,
+                                   const InputParameters& params) {
+  const uint32_t batch_size = params.num_sequences;
+  const uint32_t n_tokens = tokens.size(0);
+  const uint32_t q_max_seq_len = params.q_max_seq_len;
+  const uint32_t kv_max_seq_len = params.kv_max_seq_len;
+  const bool use_kv_cache = !params.empty_kv_cache;
+
+  // only use cuda graph for decoding batch when
+  // 1> q_max_seq_len is 1
+  // 2> max_seq_len is less than k_max_seq_len
+  // 3> batch size in the list
+  // ?? how to determine it is a decoding batch
+  if (use_kv_cache && q_max_seq_len == 1 && kv_max_seq_len <= k_max_seq_len &&
+      batch_size * q_max_seq_len == n_tokens) {
+    auto it = graphs_.find(batch_size);
+    if (it != graphs_.end()) {
+      return it->second->replay(tokens, positions, params);
+    }
+  }
+
+  return model_->forward(tokens, positions, kv_caches, params);
+}
+
+void ModelRunner::CudaGraph::capture(CausalLM* model,
+                                     torch::Tensor flatten_tokens,
+                                     torch::Tensor flatten_positions,
+                                     const InputParameters& params,
+                                     std::vector<KVCache>& kv_cache) {
+  // run model once to avoid captured graph including initial benchmarking
+  model->forward(flatten_tokens, flatten_positions, kv_cache, params);
+  torch::cuda::synchronize();
+
+  // save input tensors
+  flatten_tokens_buffer_ = flatten_tokens;
+  flatten_positions_buffer_ = flatten_positions;
+  new_cache_slots_buffer_ = params.new_cache_slots;
+  block_tables_buffer_ = params.block_tables;
+  q_cu_seq_lens_buffer_ = params.q_cu_seq_lens;
+  kv_cu_seq_lens_buffer_ = params.kv_cu_seq_lens;
+
+  // create cuda graph and capture
+  at::cuda::CUDAStream capture_stream = at::cuda::getStreamFromPool();
```

Comment:
need to create a separate stream for graph capture, otherwise the assert got fired. 
https://github.com/kulinseth/pytorch/blob/930f565a45e883d0131e9d4ba38d0ce4ca81ada4/aten/src/ATen/cuda/CUDAGraph.cpp#L79